dvc/repo/diff.py

```diff
@@ -0,0 +1,219 @@
+from __future__ import unicode_literals
+
+import os
+from errno import ENOENT
+
+
+import dvc.logger as logger
+from dvc.scm.base import FileNotInCommitError
+from dvc.scm.git import (
+    DIFF_A_TREE,
+    DIFF_B_TREE,
+    DIFF_A_REF,
+    DIFF_B_REF,
+    DIFF_EQUAL,
+)
+
+
+DIFF_TARGET = "target"
+DIFF_IS_DIR = "is_dir"
+DIFF_OLD_FILE = "old_file"
+DIFF_OLD_CHECKSUM = "old_checksum"
+DIFF_NEW_FILE = "new_file"
+DIFF_NEW_CHECKSUM = "new_checksum"
+DIFF_SIZE = "size_diff"
+DIFF_DEL = "del"
+DIFF_IDENT = "ident"
+DIFF_CHANGE = "changes"
+DIFF_NEW = "new"
+DIFF_MOVE = "moves"
+DIFF_LIST = "diffs"
+DIFF_SIZE_UNKNOWN = "?"
+DIFF_A_OUTPUT = "a_output"
+DIFF_B_OUTPUT = "b_output"
+DIFF_DELETED = "deleted_file"
+
+
+def _file_not_exists(error, result):
+    if error.errno == ENOENT:
+        result.update({DIFF_SIZE: DIFF_SIZE_UNKNOWN})
+    else:
+        raise error
+
+
+def _extract_dir(self, output):
+    """Extract the content of dvc tree file
+    Args:
+        self(object) - Repo class instance
+        output(object) - OutputLOCAL class instance
+    Returns:
+        dict - dictionary with keys - paths to file in .dvc/cache
+                               values -checksums for that files
+    """
+    lst = output.dir_cache
+    return {i["relpath"]: i["md5"] for i in lst}
+
+
+def _get_tree_changes(self, a_entries, b_entries):
+    result = {
+        DIFF_DEL: 0,
+        DIFF_IDENT: 0,
+        DIFF_CHANGE: 0,
+        DIFF_NEW: 0,
+        DIFF_MOVE: 0,
+    }
+    keys = set(a_entries.keys())
+    keys.update(b_entries.keys())
+    diff_size = 0
+    for key in keys:
+        if key in a_entries and key in b_entries:
+            if a_entries[key] == b_entries[key]:
+                result[DIFF_IDENT] += 1
+            else:
+                result[DIFF_CHANGE] += 1
+                diff_size += os.path.getsize(
+                    self.cache.local.get(b_entries[key])
+                ) - os.path.getsize(self.cache.local.get(a_entries[key]))
+        elif key in a_entries:
+            result[DIFF_DEL] += 1
+            diff_size -= os.path.getsize(self.cache.local.get(a_entries[key]))
+        else:
+            result[DIFF_NEW] += 1
+            diff_size += os.path.getsize(self.cache.local.get(b_entries[key]))
+    result[DIFF_SIZE] = diff_size
+    return result
+
+
+def _check_local_cache(a_out, b_out, is_checked):
+    if not is_checked:
+        outputs = []
+        if a_out is not None and a_out.scheme != "local":
+            is_checked = True
+            outputs.append(str(a_out))
+        if b_out is not None and b_out.scheme != "local":
+            is_checked = True
+            outputs.append(str(b_out))
+        if is_checked:
+            logger.warning("non-local output: {}".format(outputs))
```

Comment:
I think the message could be more informative. Something like "diff is not supported for non-local outputs. Ignoring '{}'"